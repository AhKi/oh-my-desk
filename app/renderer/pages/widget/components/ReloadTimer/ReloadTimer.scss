.ReloadTimer {
  display: flex;
  position: fixed;
  border-radius: 45px;
  background-color: #3C3F47;
  padding: 3px 5px;
  top: 85px;
  right: 8px;
  transition: background-color 1s;
  align-self: center;
  justify-content: center;
}

.ReloadTimer__button {
  display: flex;
  color: $color-white;

  svg{
    width: 20px;
    margin: 2px 4px 0 4px;
    fill: #a8aab1;
  }
}

.ReloadTimer__stop {
  background-color: $color-red;

  svg{
    fill: #ffffff;
  }
}

.ReloadTimer__timetext {
  line-height: 1;
  width: 50px;
}