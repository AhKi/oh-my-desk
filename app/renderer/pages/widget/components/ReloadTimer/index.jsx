import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Svg from 'react-svg-inline';
import closeIcon from 'assets/page-view_icon/icon_xbtn.svg';
import './ReloadTimer.scss';

const propTypes = {
  id: PropTypes.string,
  reloadTimer: PropTypes.number,
  webView: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  onUpdateInfo: PropTypes.func,
};
const defaultProps = {
  id: '',
  reloadTimer: 0,
  webView: null,
  onUpdateInfo() {},
};

class ReloadTimer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      timer: props.reloadTimer,
      tick: null,
    };
    this.handleCancelTimer = this.handleCancelTimer.bind(this);
    this.handleTick = this.handleTick.bind(this);
    this.handleToggleTimer = this.handleToggleTimer.bind(this);
  }

  componentDidMount() {
    this.setState({
      tick: setInterval(this.handleTick, 1000),
    });
  }

  componentDidUpdate(prevProps) {
    const { reloadTimer } = this.props;

    if (prevProps.reloadTimer !== reloadTimer) {
      this.setState({ timer: reloadTimer }); // eslint-disable-line react/no-did-update-set-state
    }
  }

  componentWillUnmount() {
    const { tick } = this.state;

    clearInterval(tick);
  }

  handleToggleTimer() {
    const { tick } = this.state;

    if (tick) {
      clearInterval(tick);
      this.setState({ tick: null });
    } else {
      this.setState({ tick: setInterval(this.handleTick, 1000) });
    }
  }

  handleTick() {
    const { reloadTimer, webView } = this.props;
    const { timer } = this.state;

    if (timer <= 0) {
      this.setState({ timer: reloadTimer });
      webView.reload();
    } else {
      this.setState(prevState => ({
        timer: prevState.timer - 1,
      }));
    }
  }

  handleCancelTimer() {
    const { id, onUpdateInfo } = this.props;

    onUpdateInfo(id, {
      reloadInterval: 0,
    });
  }

  render() {
    const { timer, tick } = this.state;
    const minute = parseInt(timer / 60, 10);
    const second = timer % 60;
    const reloadTimerBtnClassName = cx('ReloadTimer', {
      ReloadTimer__stop: !tick,
    });

    return (
      <div className={reloadTimerBtnClassName}>
        <button
          className="ReloadTimer__button"
          type="button"
          onClick={this.handleCancelTimer}
        >
          <Svg svg={closeIcon} />
        </button>
        <button
          className="ReloadTimer__button ReloadTimer__timetext"
          type="button"
          onClick={this.handleToggleTimer}
        >
          {minute < 10 ? `0${minute}` : minute} : {second < 10 ? `0${second}` : second}
        </button>
      </div>
    );
  }
}

ReloadTimer.propTypes = propTypes;
ReloadTimer.defaultProps = defaultProps;

export default ReloadTimer;
