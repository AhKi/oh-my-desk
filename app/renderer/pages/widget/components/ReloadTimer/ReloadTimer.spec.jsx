import React from 'react';
import { shallow } from 'enzyme';

import ReloadTimer from '.';

describe('<ReloadTimer />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match to snapshot when render default', () => {
    const wrapper = shallow(<ReloadTimer />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should match to snapshot when render default', () => {
    const wrapper = shallow(<ReloadTimer />);
    wrapper.setState({ timer: 6015 });

    expect(wrapper).toMatchSnapshot();
  });

  it('should call handleCancelTimer', () => {
    const onUpdateInfo = jest.fn();
    const wrapper = shallow(<ReloadTimer id="mock-id" onUpdateInfo={onUpdateInfo} />);

    wrapper.instance().handleCancelTimer();

    expect(onUpdateInfo).toHaveBeenCalledTimes(1);
    expect(onUpdateInfo).toHaveBeenCalledWith('mock-id', {
      reloadInterval: 0,
    });
  });

  describe('should call handleTick', () => {
    const webView = {
      reload: jest.fn(),
    };
    const wrapper = shallow(<ReloadTimer reloadTimer={50} webView={webView} />);

    it('when timer <= 0', () => {
      wrapper.setState({ timer: 0 });

      wrapper.instance().handleTick();

      expect(webView.reload).toHaveBeenCalledTimes(1);
      expect(webView.reload).toHaveBeenCalledWith();
      expect(wrapper.instance().state.timer).toBe(50);
    });

    it('when timer <= 0', () => {
      wrapper.setState({ timer: 10 });

      wrapper.instance().handleTick();

      expect(webView.reload).toHaveBeenCalledTimes(0);
      expect(wrapper.instance().state.timer).toBe(9);
    });
  });

  it.skip('should call componentDidUpdate', () => {
    const componentDidUpdate = jest.spyOn(ReloadTimer.prototype, 'componentDidUpdate');
    const wrapper = shallow(<ReloadTimer />);

    expect(wrapper.instance().state.timer).toBe(0);
    expect(componentDidUpdate).toHaveBeenCalledTimes(0);

    wrapper.setProps({ reloadTimer: 50 });

    expect(componentDidUpdate).toHaveBeenCalledTimes(2);
    expect(wrapper.instance().state.timer).toBe(50);
  });

  it.skip('should test componentDidMount and componentWillUnmount', () => {
    jest.useFakeTimers();
    const componentDidMount = jest.spyOn(ReloadTimer.prototype, 'componentDidMount');
    const componentWillUnmount = jest.spyOn(ReloadTimer.prototype, 'componentWillUnmount');
    const wrapper = shallow(<ReloadTimer />);
    const { handleTick, tick } = wrapper.instance();

    expect(componentDidMount).toHaveBeenCalledTimes(1);
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(setInterval).toHaveBeenCalledWith(
      handleTick,
      1000,
    );

    wrapper.unmount();
    expect(componentWillUnmount).toHaveBeenCalledTimes(1);
    expect(clearInterval).toHaveBeenCalledTimes(1);
    expect(clearInterval).toHaveBeenCalledWith(tick);
  });

  describe.skip('should test handleToggleTimer', () => {
    jest.useFakeTimers();

    const wrapper = shallow(<ReloadTimer />);

    it('when this.tick === null', () => {
      wrapper.instance().tick = null;
      wrapper.instance().handleToggleTimer();

      expect(setInterval).toHaveBeenCalledTimes(1);
      expect(setInterval).toHaveBeenCalledWith(
        wrapper.instance().handleTick,
        1000,
      );
    });

    it('when this.tick !== null', () => {
      wrapper.instance().tick = 50;
      wrapper.instance().handleToggleTimer();

      expect(clearInterval).toHaveBeenCalledTimes(1);
      expect(clearInterval).toHaveBeenCalledWith(50);
    });
  });
});
