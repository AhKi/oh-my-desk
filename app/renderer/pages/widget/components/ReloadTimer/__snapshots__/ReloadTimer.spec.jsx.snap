// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReloadTimer /> should match to snapshot when render default 1`] = `
<div
  className="ReloadTimer"
>
  <button
    className="ReloadTimer__button"
    onClick={[Function]}
    type="button"
  >
    <SVGInline
      classSuffix="-svg"
      cleanup={Array []}
      cleanupExceptions={Array []}
      component="span"
      svg={Object {}}
    />
  </button>
  <button
    className="ReloadTimer__button"
    onClick={[Function]}
    type="button"
  >
    00
     : 
    00
  </button>
</div>
`;

exports[`<ReloadTimer /> should match to snapshot when render default 2`] = `
<div
  className="ReloadTimer"
>
  <button
    className="ReloadTimer__button"
    onClick={[Function]}
    type="button"
  >
    <SVGInline
      classSuffix="-svg"
      cleanup={Array []}
      cleanupExceptions={Array []}
      component="span"
      svg={Object {}}
    />
  </button>
  <button
    className="ReloadTimer__button"
    onClick={[Function]}
    type="button"
  >
    100
     : 
    15
  </button>
</div>
`;
